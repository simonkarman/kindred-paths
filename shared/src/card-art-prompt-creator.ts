/* eslint-disable max-len */
import { Card, landSubtypeToColor } from './card';

const fixSpacing = (text: string): string => {
  return text
    .replace(/\s+/g, ' ') // Replace multiple instances of whitespace with a single space
    .trim(); // Trim leading and trailing spaces
};

export class CardArtPromptCreator {
  constructor() {}

  createPrompt(card: Card): string {
    const sections: string[] = [];

    // Add introduction section
    const dominantCardType = card.types[card.types.length - 1];
    sections.push(`A Magic the Gathering artwork for a ${dominantCardType} named "${card.name}"`);

    // Style section
    const focus: { [cardType: string]: string } = {
      'creature': ', focusing on a dynamic moment that tells a clear story within a single frame',
      'artifact': ', focusing on a close up of the artifact as the main subject',
      'sorcery': ', focusing on an event taking place',
      'instant': ', focusing on an action being performed',
      'land': ', focusing on an empty landscape or environment',
      'enchantment': ', focusing on a magical effect that is visible in the environment',
      'planeswalker': ', focusing on a dynamic moment that tells a clear story within a single frame',
    };
    if (card.tags['set'] === 'miffy') {
      const miffyStyle = `in the style of Miffy by Dick Bruna using simple colors and without shadows${focus[dominantCardType] ?? ''}.`;
      sections.push(miffyStyle);
    } else {
      const magicStyle = `in the style of Magic: The Gathering artists like Johannes Voss, Chris Rahn, or Magali Villeneuve${focus[dominantCardType] ?? ''}.`;
      sections.push(magicStyle);
    }

    // Add main subject section when the card is not a land
    if (dominantCardType !== 'land') {
      sections.push('The main subject takes up about 60% of the composition.');
    }

    // Add card type section
    sections.push(this[`${dominantCardType}Prompt`](card));

    // Add tags["setting"] section
    const setting = card.tags['setting'];
    if (typeof setting === 'string' && setting.length > 0) {
      sections.push(setting);
    }

    return sections.filter(s => s.length > 0).join(' ');
  }

  private creaturePrompt(card: Card): string {
    const power = card.pt?.power ?? 0;
    const toughness = card.pt?.toughness ?? 0;

    const powerDescription = {
      0: 'non-combatant',
      1: toughness > 1 ? 'weak' : '',
      2: '',
      3: '',
      4: 'strength',
    }[power] ?? 'powerful';

    const creatureTypeDescription = {
      0: 'creature',
      1: `${card.subtypes[0]} creature`,
      2: `${card.subtypes[0]} creature with ${card.subtypes[1]} features`,
    }[card.subtypes.length] || `${card.subtypes[0]} creature with features from ${card.subtypes.slice(1, -1).join(', ')} and ${card.subtypes[card.subtypes.length - 1]}`;

    const toughnessDescription = {
      0: 'nimble-sized',
      1: 'nimble-sized in an agile pose',
      2: 'small-sized in a dynamic pose',
      3: 'medium-sized with balanced proportions',
      4: 'medium-sized with a robust build',
      5: 'large-sized with a powerful stance',
      6: 'large-sized with overwhelming presence',
    }[toughness] || 'massive towering creature with an overwhelming presence';

    const singleKeywordDescriptions: { [keyword: string]: string } = {
      'flying': 'is flying',
      'trample': 'shows unstoppable force',
      'haste': 'is charging forward',
      'first strike': 'is ready to strike first in combat',
      'vigilance': 'is standing alert and ready for action',
      'reach': 'has a feature that allows it to reach (or strike targets) in the sky',
    };
    const keywordDescriptions = card.rules
      .filter(r => r.variant === 'keyword' && r.content in singleKeywordDescriptions)
      .map(r => singleKeywordDescriptions[r.content]);
    const keywordDescription = keywordDescriptions.length > 0 ? `which ${keywordDescriptions.join(' and ')}` : '';

    const rarityDescription = {
      'common': 'simple appearance with common and standard clothing',
      'uncommon': 'simple appearance with common and standard clothing',
      'rare': 'intricate appearance with unique and rare clothing',
      'mythic': 'mythical appearance with awe-inspiring presence and clothing made with incredible craftsmanship',
    }[card.rarity];

    const color = card.color();
    const colorDescription = color.length > 0 ? color.join(' and ') : 'gray and brown';

    return fixSpacing(
      `Depicting a ${card.supertype ?? ''} ${powerDescription} ${creatureTypeDescription} `
      + `that is a ${toughnessDescription} ${keywordDescription} use a ${rarityDescription} and focus on ${colorDescription} colors.`,
    );
  }

  private artifactPrompt(card: Card): string {
    const rarityDescription = {
      'common': 'steel look with common and standard shapes',
      'uncommon': 'metallic look with common shapes',
      'rare': 'golden look of high quality made with precious materials',
      'mythic': 'unique glowing look of incredible craftsmanship made with elaborate details and mythical materials',
    }[card.rarity];

    const color = card.colorIdentity();
    const colorDescription = color.length > 0 ? color.join(' and ') : 'gray and brown';

    return `${card.name} is a small item, that is shown in a fitting environment. It has a ${rarityDescription} and the item has ${colorDescription} colors.`;
  }

  private enchantmentPrompt(card: Card): string {
    const isAura = card.subtypes.includes('aura');
    return `Showing ${card.name} as a ${card.color().join(' and ')} magical effect ${isAura ? 'on a creature.' : 'that is visible in the environment.'}`;
  }

  private instantPrompt(card: Card): string {
    return '';
  }

  private landPrompt(card: Card): string {
    const identityColors = card.colorIdentity();
    const landTypeColors = card.subtypes.map(st => landSubtypeToColor(st)).filter(c => c !== undefined);
    const colors = Array.from(new Set([...identityColors, ...landTypeColors]));

    const subtypeDescription = card.subtypes.length > 0 ? `is a ${card.subtypes.join(' and ')} and ` : '';

    const landDescriptions = {
      white: 'vast open plains, rolling grasslands, serene horizon',
      blue: 'mystical island, flowing water, coastal beauty',
      black: 'dark swampland, murky waters, twisted vegetation',
      red: 'rugged mountain terrain, rocky peaks, volcanic elements',
      green: 'lush forest landscape, ancient trees, verdant foliage',
    };
    const landDescription = colors.map(c => landDescriptions[c]).join(' and ');

    return fixSpacing(
      `${card.name} ${subtypeDescription} is a ${landDescription}.`,
    );
  }

  private sorceryPrompt(card: Card): string {
    return '';
  }

  private planeswalkerPrompt(card: Card): string {
    return '';
  }
}
