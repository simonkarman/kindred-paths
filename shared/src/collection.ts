export type StorageBackend =
  | { type: 'none' }
  | {
  type: 'git',
  repositoryUrl: string,
  branch: string
};

export type FileModification = {
  operation: 'addition' | 'modification' | 'deletion',
  path: string,
};

export type LocalSyncStatus = {
  lastSyncCommit: string,
  lastSyncTime: string,
  modifications: FileModification[],
};

export type RemoteSyncStatus = {
  hasIncomingChanges: boolean,
  lastChecked: string,
};

export type SyncStatus = {
  local: LocalSyncStatus,
  remote: RemoteSyncStatus,
};

export type Collection = {
  storage: StorageBackend,
  status: SyncStatus
};

export type SyncResult =
  | {
  success: true,
  message: string,
}
  | {
  success: false,
  error: string,
  details?: string,
};
