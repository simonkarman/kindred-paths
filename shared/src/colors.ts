export type CardColor = 'white' | 'blue' | 'black' | 'red' | 'green';
export type Mana = CardColor | 'colorless';
export const cardColors = ['white', 'blue', 'black', 'red', 'green'] as const;
export const wubrg = ['w', 'u', 'b', 'r', 'g'] as const;
export type CardColorCharacter = typeof wubrg[number];

export const colorToShort = (color: CardColor): CardColorCharacter => {
  switch (color) {
    case 'white': return 'w';
    case 'blue': return 'u';
    case 'black': return 'b';
    case 'red': return 'r';
    case 'green': return 'g';
  }
}

export const colorToLong = (color: CardColorCharacter): CardColor => {
  switch (color) {
    case 'w': return 'white';
    case 'u': return 'blue';
    case 'b': return 'black';
    case 'r': return 'red';
    case 'g': return 'green';
    default: throw new Error(`Unknown color: ${color}`);
  }
}

export const toOrderedColors = (_colors: CardColor[]): CardColor[] => {
  // Break early for empty or single color arrays
  if (_colors.length === 0) return [];
  if (_colors.length === 1) return [..._colors];

  // Check for duplicates
  const uniqueColors = new Set(_colors);
  if (uniqueColors.size !== _colors.length) {
    throw new Error(`duplicate colors found in: [${_colors.join(', ')}]`);
  }

  // Check that each color in set is a valid card color
  for (const color of uniqueColors) {
    if (!cardColors.includes(color)) {
      throw new Error(`invalid color "${color}" found in: [${_colors.join(', ')}]`);
    }
  }
  const wubrgOrdered = cardColors.filter(color => uniqueColors.has(color));

  // If two colors are present, ...
  if (wubrgOrdered.length === 2) {
    const indexA = cardColors.indexOf(wubrgOrdered[0]);
    const indexB = cardColors.indexOf(wubrgOrdered[1]);

    // If their neighboring colors are present, return them in WUBRG order, and for green/white return them in reverse order
    if ((indexB - indexA) === 1) {
      return wubrgOrdered;
    }
    if (indexA === 0 && indexB === 4) {
      return wubrgOrdered.reverse();
    }

    // If the two colors are not neighboring, return them in WUBRG order, except for white/red and blue/green which should be reversed
    if ((indexB - indexA) === 2) {
      return wubrgOrdered;
    }
    if ((indexA === 0 && indexB === 3) || (indexA === 1 && indexB === 4)) {
      return wubrgOrdered.reverse();
    }
  }

  // If three colors are present, ...
  if (wubrgOrdered.length === 3) {
    const indexA = cardColors.indexOf(wubrgOrdered[0]);
    const indexB = cardColors.indexOf(wubrgOrdered[1]);
    const indexC = cardColors.indexOf(wubrgOrdered[2]);

    // If the colors are in a sequence, return them in WUBRG order
    if ((indexB - indexA) === 1 && (indexC - indexB) === 1) {
      return wubrgOrdered;
    }
    if ((indexA === 0 && indexB === 1 && indexC === 4)) {
      return ['green', 'white', 'blue'];
    }
    if ((indexA === 0 && indexB === 3 && indexC === 4)) {
      return ['red', 'green', 'white'];
    }

    // Otherwise, find the color that is not neighboring one of the others
    const [missingIndexA, missingIndexB] = [0, 1, 2, 3, 4].filter(colorIndex => !wubrgOrdered.includes(cardColors[colorIndex]));
    const enemyColorIndex = (missingIndexB - missingIndexA) === 2
      ? missingIndexA + 1
      : (missingIndexB + 1) % cardColors.length;

    // And return the enemy color in the middle of the other two colors
    return [
      cardColors[(enemyColorIndex + 3) % cardColors.length],
      cardColors[enemyColorIndex],
      cardColors[(enemyColorIndex + 2) % cardColors.length]
    ]
  }

  // If four colors are present, return them in WUBRG order starting the loop after the missing color
  if (wubrgOrdered.length === 4) {
    const missingColor = cardColors.find(color => !uniqueColors.has(color));
    if (!missingColor) throw new Error(`missing color not found in four-color set: [${_colors.join(', ')}]`);
    const startIndex = (cardColors.indexOf(missingColor) + 1) % cardColors.length;
    return [
      ...cardColors.slice(startIndex, startIndex + 4),
      ...cardColors.slice(0, Math.max(0, startIndex - 1))
    ];
  }

  // If all colors are present, return them in WUBRG order
  if (_colors.length === 5) {
    return wubrgOrdered;
  }

  // This should never happen
  throw new Error(`unhandled case in toOrderedColors for [${_colors.join(', ')}]`);
}
