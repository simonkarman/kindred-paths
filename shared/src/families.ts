import { CardColor } from './card';

export type Family = {
  name: string,
  color: {
    primary: CardColor,
    secondary: CardColor,
  },
  focus: 'legacy' | 'future',
  race: {
    primary: string,
    secondary: string,
  },
  class: {
    primary: string,
    secondary: string,
  },
};

export const families: () => ReadonlyArray<Family> = () => [
  {
    name: 'Azurian',
    color: { primary: 'white', secondary: 'blue' },
    focus: 'future',
    race: { primary: 'fox', secondary: 'angel' },
    class: { primary: 'advisor', secondary: 'cleric' },
  },
  {
    name: 'Selenworth',
    color: { primary: 'white', secondary: 'green' },
    focus: 'future',
    race: { primary: 'cat', secondary: 'angel' },
    class: { primary: 'scout', secondary: 'cleric' },
  },
  {
    name: 'Deepmir',
    color: { primary: 'blue', secondary: 'black' },
    focus: 'legacy',
    race: { primary: 'whale', secondary: 'sphinx' },
    class: { primary: 'rogue', secondary: 'wizard' },
  },
  {
    name: 'Simovian',
    color: { primary: 'blue', secondary: 'green' },
    focus: 'legacy',
    race: { primary: 'turtle', secondary: 'sphinx' },
    class: { primary: 'elder', secondary: 'wizard' },
  },
  {
    name: 'Orzim',
    color: { primary: 'black', secondary: 'white' },
    focus: 'legacy',
    race: { primary: 'bat', secondary: 'demon' },
    class: { primary: 'knight', secondary: 'warlock' },
  },
  {
    name: 'Rakdovale',
    color: { primary: 'black', secondary: 'red' },
    focus: 'future',
    race: { primary: 'scorpion', secondary: 'demon' },
    class: { primary: 'mercenary', secondary: 'warlock' },
  },
  {
    name: 'Borheim',
    color: { primary: 'red', secondary: 'white' },
    focus: 'future',
    race: { primary: 'goat', secondary: 'dragon' },
    class: { primary: 'berserker', secondary: 'shaman' },
  },
  {
    name: 'Izzuhart',
    color: { primary: 'red', secondary: 'blue' },
    focus: 'legacy',
    race: { primary: 'phoenix', secondary: 'dragon' },
    class: { primary: 'artificer', secondary: 'shaman' },
  },
  {
    name: 'Golgathorne',
    color: { primary: 'green', secondary: 'black' },
    focus: 'legacy',
    race: { primary: 'wolf', secondary: 'hydra' },
    class: { primary: 'archer', secondary: 'druid' },
  },
  {
    name: 'Grulest',
    color: { primary: 'green', secondary: 'red' },
    focus: 'future',
    race: { primary: 'boar', secondary: 'hydra' },
    class: { primary: 'smith', secondary: 'druid' },
  },
];

export const primaryRacesPerPrimaryColor = families().reduce((acc, family) => {
  acc[family.color.primary] = (acc[family.color.primary] || []).concat(family.race.primary);
  return acc;
}, {} as Record<CardColor, string[]>);

export const primaryRacesPerAllColors = families().reduce((acc, family) => {
  acc[family.color.primary] = Array.from(new Set((acc[family.color.primary] || []).concat(family.race.primary)));
  acc[family.color.secondary] = Array.from(new Set((acc[family.color.secondary] || []).concat(family.race.primary)));
  return acc;
}, {} as Record<CardColor, string[]>);
