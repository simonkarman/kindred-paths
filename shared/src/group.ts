export type Predicate<T> = (item: T) => boolean;

export type Requirement<T> = {
  name: string;
  predicate: Predicate<T>;
}

/**
 * Represents a match where an item is successfully matched to a requirement.
 */
type Succeeded = {
  requirement: string;
  status: 'succeeded';
  id: string;
};

/**
 * Represents a requirement that failed to match any item.
 */
type Failed = {
  requirement: string;
  status: 'failed';
};

/**
 * Represents an item that is valid to at least one requirement, but has not been selected for any of those requirements.
 * This indicates that this item was ambiguously not chosen, and could replace the card for any of the requirements it matched.
 */
type Ambiguous = {
  id: string;
  status: 'ambiguous';
  requirements: string[];
};

type Match = Succeeded | Failed | Ambiguous;

export class Group<T extends { id: string }> {
  constructor(
    public readonly name: string,
    public readonly requirements: Requirement<T>[],
  ) {
    // Validate no duplicate requirement names
    const requirementNames = new Set<string>();
    for (const requirement of requirements) {
      if (requirementNames.has(requirement.name)) {
        throw new Error(`group ${name} has a duplicate requirement ${requirement.name}`);
      }
      requirementNames.add(requirement.name);
    }
  }

  public matchTo(_set: T[]): Match[] {
    // Copy the set to avoid mutating the original
    const set: Readonly<T>[] = [..._set];
    const candidates = new Set<string>();

    // Keep track of open requirements
    let open: Readonly<Requirement<T>>[] = [...this.requirements];

    // While there are still requirements left in the open set, try to find matches
    const matches: Match[] = [];
    while (open.length > 0) {

      // Find all matches per requirement
      const matchesPerRequirement: Record<string, T[]> = {};
      for (const requirement of open) {
        matchesPerRequirement[requirement.name] = set.filter(item => {
          if (requirement.predicate(item)) {
            candidates.add(item.id);
            return true;
          }
          return false;
        });
      }

      // Find the requirement with the least matches
      const requirement = Object.entries(matchesPerRequirement)
        .reduce((acc, [name, { length: numberOfMatches }]) => {
          if (numberOfMatches < acc.numberOfMatches) {
            return { name, numberOfMatches };
          }
          return acc;
        }, { name: undefined as (string | undefined), numberOfMatches: Infinity }).name;
      if (requirement === undefined) {
        throw new Error('this should never happen');
      }

      // Remove the matched requirement from the open set
      open = open.filter(req => req.name !== requirement);

      // Find the item
      const item = matchesPerRequirement[requirement].pop();
      if (item === undefined) {
        // If there is no match, mark the requirement as failed
        matches.push({
          requirement,
          status: 'failed',
        });
        continue;
      }

      // Remove the matched item from the set and from allCandidates
      set.splice(set.indexOf(item), 1);
      candidates.delete(item.id);

      // Add the match to the results
      matches.push({
        requirement,
        status: 'succeeded',
        id: item.id,
      });
    }

    if (candidates.size > 0) {
      // If there are still candidates left, they are not part of the group, but could have ambiguously been chosen for the requirements
      //  they matched for.
      for (const candidateId of candidates) {
        matches.push({
          id: candidateId,
          status: 'ambiguous',
          requirements: this.requirements.filter(req => set.some(item => item.id === candidateId && req.predicate(item))).map(r => r.name),
        });
      }
    }

    return matches;
  }
}
