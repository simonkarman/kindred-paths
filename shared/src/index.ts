export * from './card';
export * from './example-groups';
export * from './families';
export * from './group';
export * from './hash';
export * from './kindred-paths-groups';
export * from './random';
export * from './serialized-card';
