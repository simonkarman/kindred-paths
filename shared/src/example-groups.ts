import { Group } from './group';
import { Card } from './card';

export const exampleGroup1 = new Group<Card>("Example Group 1", [
  { name: 'artifact', predicate: c => c.types.includes('artifact') },
  { name: 'wurbg card', predicate: c => c.color.length === 5 },
  { name: 'mono green mythic', predicate: c => c.color.length === 1 && c.color()[0] === 'green' && c.rarity === 'mythic' },
  { name: 'legendary creature 1/2', predicate: c => c.types.includes('creature') && c.supertype === 'legendary' },
  { name: 'legendary creature 2/2', predicate: c => c.types.includes('creature') && c.supertype === 'legendary' },
  { name: 'white legendary creature', predicate: c => c.color.length === 1 && c.color().includes('white') && c.types.includes('creature') && c.supertype === 'legendary' },
]);

export const exampleGroup2 = new Group<Card>("Example Group 2", [
  { name: 'artifact 1/2', predicate: c => c.types.includes('artifact') },
  { name: 'artifact 2/2', predicate: c => c.types.includes('artifact') },
]);

export const exampleGroup3 = new Group<Card>("Example Group 3", [
  { name: 'artifact', predicate: c => c.types.includes('artifact') },
]);
