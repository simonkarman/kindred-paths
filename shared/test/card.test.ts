/* eslint-disable max-len */
import { expect, test } from 'vitest';
import { Card, landSubtypeToColor, SerializedCard, tryParseLoyaltyAbility } from '../src';

test('land subtypes', () => {
  expect(landSubtypeToColor('forest')).toBe('green');
  expect(landSubtypeToColor('plains')).toBe('white');
  expect(landSubtypeToColor('island')).toBe('blue');
  expect(landSubtypeToColor('swamp')).toBe('black');
  expect(landSubtypeToColor('mountain')).toBe('red');
  expect(landSubtypeToColor('abc')).toBeUndefined();
});

test('loyalty', () => {
  expect(tryParseLoyaltyAbility({ variant: 'keyword', content: 'flying' })).toStrictEqual({ success: false });
  expect(tryParseLoyaltyAbility({ variant: 'ability', content: '+1: Hello!' })).toStrictEqual({
    success: true, cost: 1, content: 'Hello!',
  });
});

test('loyalty - complex', () => {
  const scenarios = [
    'adflk: adfadsf',
    '3: adf',
    '+4: adf',
    '-2: adf',
    '-0: adf',
    '0: adf',
    '5: 3fad:',
    'adf6: adf',
    '9ad: 2:',
    'a: 2: adfadafsfasdf',
    '+0: adf',
    '+69: adf',
    '-96: adfasd',
    ': adf',
    '7: 323:',
    '+8:+23:',
    '+93:adsf',
    '-94:def6',
    '+11',
    '-10',
    '+10: yes',
    '+01: no',
    '-: adf',
    '+: def',
    '+13 adf',
    '-14 def',
    '+13: adfasdf jh',
    ':32: adf',
    '+-adf: adf',
    '+15: A text that is longer with multile spaces (weird characters)',
    '+16adfadf',
    '+17 adf',
    '-X: adfasdf',
    '+X: yes',
    'X: no',
    'nno: +X',
    '-x: no',
    '+x: no',
  ];
  expect(scenarios.map(s => tryParseLoyaltyAbility({ variant: 'ability', content: s }))).toMatchSnapshot();
});

test('creature card', () => {
  const serializedCreatureCard: SerializedCard = {
    id: 'creature',
    name: 'Sam, The Great',
    rarity: 'rare',
    isToken: undefined,
    supertype: undefined,
    tokenColors: undefined,
    types: ['creature'],
    subtypes: ['human', 'warrior'],
    manaCost: { x: 2, colorless: 2, white: 1, blue: 0 },
    rules: [
      { variant: 'card-type-reminder', content: 'A creature can\'t attack unless it has been under its controller\'s control continuously since their most recent turn began.' },
      { variant: 'keyword', content: 'vigilance' },
      { variant: 'keyword', content: 'first strike' },
      { variant: 'keyword', content: 'lifelink' },
      { variant: 'inline-reminder', content: 'Damage dealt by this creature also causes you to gain that much life.' },
      { variant: 'keyword', content: 'double strike' },
      { variant: 'keyword', content: 'flying' },
      { variant: 'keyword', content: 'extort' },
      { variant: 'inline-reminder', content: 'Whenever you cast a spell, you may pay {w/b}. If you do, each opponent loses 1 life and you gain that much life.' },
      { variant: 'ability', content: 'When ~ enters, add {g} or {r/w}.' },
      { variant: 'flavor', content: 'To be, or not to be, that is the question.' },
    ],
    pt: { power: 2, toughness: 2 },
    loyalty: undefined,
    collectorNumber: 1,
    art: undefined,
    tags: { 'test': true },
  };
  const creatureCard = new Card(serializedCreatureCard);
  expect(creatureCard.manaValue()).toBe(3);
  expect(creatureCard.renderManaCost()).toBe('{x}{x}{2}{w}');
  expect(creatureCard.renderTypeLine()).toBe('Creature — Human Warrior');
  expect(creatureCard.renderRules()).toBe('{i}(A creature can\'t attack unless it has been under its controller\'s control continuously since their most recent turn began.){/i}\n' +
    'Vigilance, first strike\n' +
    'Lifelink {i}(Damage dealt by this creature also causes you to gain that much life.){/i}\n' +
    'Double strike, flying\n' +
    'Extort {i}(Whenever you cast a spell, you may pay {w/b}. If you do, each opponent loses 1 life and you gain that much life.){/i}\n' +
    'When Sam enters, add {g} or {r/w}.{flavor}To be, or not to be, that is the question.',
  );
  expect(creatureCard.loyaltyAbilities()).toStrictEqual([]);
  expect(creatureCard.color()).toStrictEqual(['white']);
  expect(creatureCard.colorIdentity()).toStrictEqual(['red', 'green', 'white']);
  expect(creatureCard.toJson()).toStrictEqual(serializedCreatureCard);
  expect(creatureCard.getReferenceName()).toBe('2/2 white Human Warrior creature');
  expect(creatureCard.explain()).toBe('"Sam, The Great" (#1) is a rare 2/2 white Human Warrior creature for {x}{x}{2}{w} mana, with: "vigilance", "first strike", "lifelink", "double strike", "flying", "extort" and "When ~ enters, add {g} or {r/w}.".');
  expect(creatureCard.getCreatableTokenNames()).toStrictEqual([]);
  expect(creatureCard.getTagAsString('test')).toBe(undefined);
  expect(creatureCard.getTagAsString('test', { stringify: true })).toBe('true');
  expect(creatureCard.getTagAsNumber('test')).toBe(undefined);
  expect(creatureCard.getTagAsNumber('test', { parseFromBoolean: true })).toBe(1);
});

test('token card', () => {
  const serializedTokenCard: SerializedCard = {
    id: '821-1-1-white-mouse-token',
    'name': 'Mouse',
    'rarity': 'common',
    'isToken': true,
    'tokenColors': [
      'white',
    ],
    'types': [
      'creature',
    ],
    'subtypes': [
      'mouse',
    ],
    'manaCost': {},
    'rules': [],
    'pt': {
      'power': 1,
      'toughness': 1,
    },
    'collectorNumber': 821,
    'art': '821-1-1-white-mouse-token-1495ce55-c5f9-4f25-bcf0-9957f4c3d0d6.png',
    'tags': {
      'status': 'concept',
      'createdAt': '2025-09-01',
      'setting': 'A small gray mouse stands triumphantly atop a wheel of cheese in a moonlit pantry, crumbs scattered around its tiny paws.',
    },
  };
  const tokenCard = new Card(serializedTokenCard);
  expect(tokenCard.manaValue()).toBe(0);
  expect(tokenCard.renderManaCost()).toBe('');
  expect(tokenCard.renderTypeLine()).toBe('Token Creature — Mouse');
  expect(tokenCard.renderRules()).toBe('');
  expect(tokenCard.loyaltyAbilities()).toStrictEqual([]);
  expect(tokenCard.color()).toStrictEqual(['white']);
  expect(tokenCard.colorIdentity()).toStrictEqual(['white']);
  expect(tokenCard.toJson()).toStrictEqual({ ...serializedTokenCard, supertype: undefined, loyalty: undefined });
  expect(tokenCard.getReferenceName()).toBe('1/1 white Mouse creature token');
  expect(tokenCard.explain()).toBe('"Mouse" (#821) is a common 1/1 white Mouse creature token.');
  expect(tokenCard.getCreatableTokenNames()).toStrictEqual([]);
  expect(tokenCard.getTagAsString('status')).toBe('concept');
});

test('planeswalker card', () => {
  const serializedPlaneswalkerCard: SerializedCard = {
    id: 'farock-the-damned-doctor',
    'name': 'Farock, The Damned Doctor',
    'rarity': 'rare',
    'supertype': 'legendary',
    'types': [
      'planeswalker',
    ],
    'subtypes': [],
    'manaCost': {
      'colorless': 1,
      'black': 2,
    },
    'rules': [
      {
        'variant': 'keyword',
        'content': 'deathtouch',
      },
      {
        'variant': 'keyword',
        'content': 'lifelink',
      },
      {
        variant: 'ability',
        content: 'When ~ enters, you gain 2 life.',
      },
      {
        'variant': 'ability',
        'content': '+1: Create a 1/1 black Zombie creature token with menace.',
      },
      {
        'variant': 'ability',
        'content': '-4: Deal 1 damage to any target. Create a Treasure token.',
      },
      {
        'variant': 'ability',
        'content': '-X: Deal 1 damage to up to X targets.',
      },
    ],
    'loyalty': 3,
    'collectorNumber': 822,
  };
  const planeswalkerCard = new Card(serializedPlaneswalkerCard);
  expect(planeswalkerCard.manaValue()).toBe(3);
  expect(planeswalkerCard.renderManaCost()).toBe('{1}{b}{b}');
  expect(planeswalkerCard.renderTypeLine()).toBe('Legendary Planeswalker — Farock');
  expect(planeswalkerCard.renderRules()).toBe('Deathtouch, lifelink\nWhen Farock enters, you gain 2 life.');
  expect(planeswalkerCard.loyaltyAbilities()).toStrictEqual([
    { cost: 1, content: 'Create a 1/1 black Zombie creature token with menace.' },
    { cost: -4, content: 'Deal 1 damage to any target. Create a Treasure token.' },
    { cost: '-X', content: 'Deal 1 damage to up to X targets.' },
  ]);
  expect(planeswalkerCard.color()).toStrictEqual(['black']);
  expect(planeswalkerCard.colorIdentity()).toStrictEqual(['black']);
  expect(planeswalkerCard.toJson()).toStrictEqual({ ...serializedPlaneswalkerCard, art: undefined, pt: undefined, tags: {}, tokenColors: undefined, isToken: undefined });
  expect(planeswalkerCard.getReferenceName()).toBe('legendary black planeswalker');
  expect(planeswalkerCard.explain()).toBe('"Farock, The Damned Doctor" (#822) is a rare legendary black planeswalker for {1}{b}{b} mana with 3 starting loyalty, with: "deathtouch", "lifelink" and "When ~ enters, you gain 2 life." and "+1: Create a 1/1 black Zombie creature token with menace." and "-4: Deal 1 damage to any target. Create a Treasure token." and "-X: Deal 1 damage to up to X targets.".');
  expect(planeswalkerCard.getCreatableTokenNames()).toStrictEqual(['1/1 black Zombie creature token with menace', 'Treasure token']);
});
