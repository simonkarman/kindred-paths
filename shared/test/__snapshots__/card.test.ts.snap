// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`loyalty - complex 1`] = `
[
  {
    "success": false,
  },
  {
    "success": false,
  },
  {
    "content": "adf",
    "cost": 4,
    "success": true,
  },
  {
    "content": "adf",
    "cost": -2,
    "success": true,
  },
  {
    "success": false,
  },
  {
    "content": "adf",
    "cost": 0,
    "success": true,
  },
  {
    "success": false,
  },
  {
    "success": false,
  },
  {
    "success": false,
  },
  {
    "success": false,
  },
  {
    "success": false,
  },
  {
    "content": "adf",
    "cost": 69,
    "success": true,
  },
  {
    "content": "adfasd",
    "cost": -96,
    "success": true,
  },
  {
    "success": false,
  },
  {
    "success": false,
  },
  {
    "success": false,
  },
  {
    "success": false,
  },
  {
    "success": false,
  },
  {
    "success": false,
  },
  {
    "success": false,
  },
  {
    "content": "yes",
    "cost": 10,
    "success": true,
  },
  {
    "success": false,
  },
  {
    "success": false,
  },
  {
    "success": false,
  },
  {
    "success": false,
  },
  {
    "success": false,
  },
  {
    "content": "adfasdf jh",
    "cost": 13,
    "success": true,
  },
  {
    "success": false,
  },
  {
    "success": false,
  },
  {
    "content": "A text that is longer with multile spaces (weird characters)",
    "cost": 15,
    "success": true,
  },
  {
    "success": false,
  },
  {
    "success": false,
  },
  {
    "content": "adfasdf",
    "cost": "-X",
    "success": true,
  },
  {
    "success": false,
  },
  {
    "success": false,
  },
  {
    "success": false,
  },
  {
    "success": false,
  },
  {
    "success": false,
  },
]
`;
